WordPress/gutenberg
# Backport r57868 from WordPress-Develop.

## What?

Backports [r57868](https://core.trac.wordpress.org/changeset/57868) / https://github.com/WordPress/wordpress-develop/commit/ca8d78e5f2f50c59749398b5932e246ada0c0a37 to the plugin's 6.5 compatibility shims.

Fixes #60140.

The WordPress-Develop commit message is:

```
Editor: Prevent font folder naive filtering causing infinite loops.

This modifies the font directory API to more closely reflect the upload directory API to help account for naive filtering when uploading fonts.

This moves the protection of infinite loops to the new function `_wp_filter_font_directory()` to allow developers extending and maintaining the font library to apply the filter without the need for a closure.

These changes also ensure both the `upload_dir` and `font_dir` filter are applied consistently when both creating and deleting fonts faces. Prior to this commit the `upload_dir` filter was only fired when creating fonts faces via the REST API.

Applying the font directory filter to the `upload_dir` filter is now done by adding the `_wp_filter_font_directory` function rather than `wp_get_font_dir()`. Developers who have previously modified the font upload directory using the `font_dir` filter will NOT need to upload their code.

Extenders wishing to upload files to the font directory can do so via the code:

{{{#!php
<?php
add_filter( 'upload_dir', '_wp_filter_font_directory' );
// Your code to upload or sideload a font file.
remove_filter( 'upload_dir', '_wp_filter_font_directory' );
}}}

Introduces:

* `wp_font_dir()`: Attempt to create and retrieve the font upload directory. The equivalent to `wp_upload_dir()`.
* `_wp_filter_font_directory()`: To run on the `upload_dir` filter, this sets the default destination of the fonts directory and fires the `font_dir` filter. 

`wp_get_font_dir()` has been modified to be a lightweight getter for the font directory. It returns the location without attempting to create it. The equivalent to `wp_get_upload_dir()`.

Follow up to [57740].

Props peterwilsoncc, mukesh27, mikachan, costdev, mmaattiiaass, swissspidy, youknowriad, dd32, grantmkin.
Fixes https://core.trac.wordpress.org/ticket/60652.
```

## Why?

Consistency with WordPress Core in future version of the plugins.


## How?

Copy paste mainly, please review the changes to make sure I haven't messed anything up

## Testing Instructions
1. Compare the copy-paste to make sure I haven't missed anything. 
2. Install WordPress 6.4.x
3. Activate Gutenberg
4. Ensure it doesn't trigger a fatal error
5. Ensure uploading a font from your computer works
6. Ensure installing a font from Google works
7. Install WordPress 6.5 (trunk)
8. Ensure it doesn't trigger a fatal error
9. Ensure uploading a font from your computer works
10. Ensure installing a font from Google works
11. Install the [Fonts To Uploads plugin](https://wordpress.org/plugins/fonts-to-uploads/)
12. Install another font via Google
13. Ensure the filter continues to work as expected and fonts are uploaded to the new destination

### Testing Instructions for Keyboard
N/A

## Screenshots or screencast 
N/A
